multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/MachineLoginFilter.java

```diff
@@ -80,75 +74,46 @@ public void init(FilterConfig filterConfig) {}
   @Override
   public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
       throws IOException, ServletException {
-    final HttpServletRequest httpRequest = (HttpServletRequest) request;
-    final String token = tokenExtractor.getToken(httpRequest);
+
+    final String token = tokenExtractor.getToken((HttpServletRequest) request);
     if (isNullOrEmpty(token)) {
       chain.doFilter(request, response);
       return;
     }
-    // check token signature and verify is this token machine or not
     try {
-      HttpSession session = ((HttpServletRequest) request).getSession(true);
-      Subject sessionSubject = (Subject) session.getAttribute("che_subject");
-      if (sessionSubject == null || !sessionSubject.getToken().equals(token)) {
-        try {
-          sessionSubject = extractSubject(token);
-          session.setAttribute("che_subject", sessionSubject);
-        } catch (NotFoundException e) {
-          sendErr(
-              response,
-              SC_UNAUTHORIZED,
-              "Authentication with machine token failed because user for this token no longer exist.");
-          return;
-        }
-      }
-
-      try {
-        EnvironmentContext.getCurrent().setSubject(sessionSubject);
-        chain.doFilter(addUserInRequest(httpRequest, sessionSubject), response);
-      } finally {
-        EnvironmentContext.reset();
-      }
-    } catch (NotMachineTokenJwtException ex) {
-      // not a machine token, bypass
+      super.doFilter(request, response, chain);
+    } catch (NotMachineTokenJwtException e) {
+      // bypass - not our token
       chain.doFilter(request, response);
-    } catch (ServerException | JwtException e) {
-      sendErr(
+    } catch (JwtException e) {
+      sendError(
           response,
           SC_UNAUTHORIZED,
           format("Authentication with machine token failed cause: %s", e.getMessage()));
     }
   }
 
-  private Subject extractSubject(String token) throws NotFoundException, ServerException {
-    final Claims claims = jwtParser.parseClaimsJws(token).getBody();
-    final String userId = claims.get(USER_ID_CLAIM, String.class);
-    // check if user with such id exists
-    final String userName = userManager.getById(userId).getName();
-    final String workspaceId = claims.get(WORKSPACE_ID_CLAIM, String.class);
-    return new MachineTokenAuthorizedSubject(
-        new SubjectImpl(userName, userId, token, false), permissionChecker, workspaceId);
-  }
-
-  /** Sets given error code with err message into give response. */
-  private static void sendErr(ServletResponse res, int errCode, String msg) throws IOException {
-    final HttpServletResponse response = (HttpServletResponse) res;
-    response.sendError(errCode, msg);
+  @Override
+  public Subject extractSubject(String token) {
+    try {
```

Comment:
this block looks NPE fragile. Do you think it would make sense to catch NPE and rethrow with some meaningful message? Or check potential NPE places for null ?

Reply:
`null` tokens are handled in `handleMissingToken()`